ddsketch/src/lib.rs

```diff
@@ -0,0 +1,461 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2023-Present
+// Datadog, Inc.
+
+//! This crate defines a minimal implementation of DDSketch.
+//!
+//! DDSketch is a data sketch used to generate percentiles over streaming data using constant
+//! memory. A DDSketch is essentially a histogram that partitions the range of positive values into
+//! an infinite number of indexed bins whose size grows exponentially. It keeps track of the number
+//! of values (or possibly floating-point weights) added to each bin. Negative values are
+//! partitioned like positive values, symmetrically to zero. The value zero as well as its close
+//! neighborhood that would be mapped to extreme bin indexes is mapped to a specific counter.
+use std::collections::{HashMap, VecDeque};
+
+use prost::Message;
+
+/// Protobuf representation of DDSketch - generated by build.rs
+pub mod pb;
+
+/// This is a minimal DDSketch implementation
+///
+/// This implementation only supports a part of the standard (which is also only the parts dd
+/// backend supports :shrug:)
+/// - max length contiguous bin store, with lower bin
+/// collapse behavior.
+/// - Positive or zero values
+///
+/// The default sketch has a 1% relative accuracy, and only accepts positive points
+///
+/// See <https://github.com/DataDog/sketches-go> for the reference implementation
+#[derive(Debug, Default, Clone)]
+pub struct DDSketch {
+    store: LowCollapsingDenseStore, // Store the weight of each bin
+    zero_count: f64,                // Store the weight of the bin of value 0
+    mapping: LogMapping,            // Bin-Value mapping
+}
+
+impl DDSketch {
+    /// Return an iterator over `(value, weight)` pair for each bin
+    pub fn ordered_bins(&self) -> Vec<(f64, f64)> {
+        let mut bins: Vec<_> = std::iter::once((0.0, self.zero_count))
+            .chain(self.store.bins().map(|(b, v)| (self.mapping.value(b), v)))
+            .collect();
+        bins.sort_by(|a, b| a.0.total_cmp(&b.0));
+        bins
+    }
+
+    /// Add a point with value `point` to the sketch
+    pub fn add(&mut self, point: f64) -> Option<()> {
+        self.add_with_count(point, 1.0)
+    }
+
+    /// Add `count` point with value `point` to the sketch
+    pub fn add_with_count(&mut self, point: f64, count: f64) -> Option<()> {
```

Comment:
Is there a particular reason that an empty Option was used as the return type? It feels non-idiomatic to me. If I were a caller of this function and the add didn't succeed because I passed in invalid inputs I'd expect a `Result<(), Error>` and for the error to explain why (even if it's just a generic "invalid input" error).